plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,908 @@
+		return fmt.Errorf("unable to create V2 network client %v", err)
+	}
+
+	// The Orchestration service is optional
+	if o.ContainsService("orchestration") {
+		if o.stack, err = openstack.NewOrchestrationV1(provider, gophercloud.EndpointOpts{}); err != nil {
+			return fmt.Errorf("unable to create V1 stack client %v", err)
+		}
+	}
+
+	// The Cinder volume storage service is optional
+	if o.ContainsService("volumev2") {
+		if o.volume, err = openstack.NewBlockStorageV2(provider, gophercloud.EndpointOpts{}); err != nil {
+			return fmt.Errorf("unable to create V2 volume client %v", err)
+		}
+	}
+
+	return nil
+}
+
+// Gather gathers resources from the OpenStack API and accumulates metrics.  This
+// implements the Input interface.
+func (o *OpenStack) Gather(acc telegraf.Accumulator) error {
+	// Gather resources.  Note service harvesting must come first as the other
+	// gatherers are dependant on this information.
+	gatherers := map[string]func(telegraf.Accumulator) error{
+		"projects":      o.gatherProjects,
+		"hypervisors":   o.gatherHypervisors,
+		"flavors":       o.gatherFlavors,
+		"servers":       o.gatherServers,
+		"volumes":       o.gatherVolumes,
+		"storage_pools": o.gatherStoragePools,
+		"subnets":       o.gatherSubnets,
+		"ports":         o.gatherPorts,
+		"networks":      o.gatherNetworks,
+		"aggregates":    o.gatherAggregates,
+		"nova_services": o.gatherNovaServices,
+		"agents":        o.gatherAgents,
+		"stacks":        o.gatherStacks,
+	}
+
+	for _, service := range o.EnabledServices {
+		if service == "services" {
+			o.accumulateServices(acc)
+			continue
+		}
+		gatherer := gatherers[service]
+		if err := gatherer(acc); err != nil {
+			acc.AddError(fmt.Errorf("failed to get resource %q %v", service, err))
+		}
+	}
+
+	if o.ServerDiagnotics && !o.InEnabledServices("servers") {
+		if err := o.gatherServers(acc); err != nil {
+			acc.AddError(fmt.Errorf("failed to get resource server diagnostics %v", err))
+		} else {
+			o.accumulateServerDiagnostics(acc)
+		}
+	}
+
+	return nil
+}
+
+// gatherStacks collects stacks from the OpenStack API.
+func (o *OpenStack) gatherStacks(acc telegraf.Accumulator) error {
+	page, err := stacks.List(o.stack, &stacks.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list stacks %v", err)
+	}
+	stacks, err := stacks.ExtractStacks(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract stacks %v", err)
+	}
+	for _, stack := range stacks {
+		tags := map[string]string{
+			"description": stack.Description,
+			"name":        stack.Name,
+			"stack_tags":  strings.Join(stack.Tags[:], ","),
+		}
+		fields := map[string]interface{}{
+			"status":        stack.Status,
+			"id":            stack.ID,
+			"status_reason": stack.StatusReason,
+			"creation_time": stack.CreationTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
+			"updated_time":  stack.UpdatedTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
```

Comment:
It's a matter of taste if you want to store the times as `string` (as you do) or as a nanosecond `int64` timestamp. If you want to do computations, the `int64` is easier on downstream queries, however the `string` is easier to read for humans. How about adding a user-configurable flag that let's you decide the format e.g. `timestamp_format = "unix_ns"`. Your format string could be the default...

Reply:
I agree, but wouldn't that create a problem in InfluxDB as `unix_ns` is `int64` format and human readable is in `string` (to different formats for same field/tag)?
And storing `unix_ns` as `string` wouldn't add any value, so I think there is only one possibility if `unix_ns` is needed, that is to store time in both formats, in two different fields.
Thoughts?

Reply:
What I meant was that the user can decide if he/she wants a human readable format or a machine readable format in influxdb. Currently, you get a "human" readable format and, if you want to do computations with it, you need to convert it in your application.

Reply:
Sorry, but I still do not see an issue here. Can you please help me, why you see a problem in creating e.g. a _field_ called `creation_time` and let the user decide if that field contains either a _string_ (human readable time as above) or a _number_ (machine readable time)? If you _know_ what "format" `creation_time` is in (e.g. `unix_ns`), then the option could be `human_readable_timestamps = true/false`. Where do you see an issue?

Reply:
The issue is if a user initially configured say `unix_ns` (default) and later chose to switch to human readable format then I think the user would not be able to store data in same DB as `creation_time` field would be of type `int` but the new data has `creation_time` field as type `string` (human readable).

Reply:
Well then the user can still create a new series by adding a tag or by changing the metric name. I think we should not work around these kind of issues in telegraf. You can still explicitly change the field name using a processor if you need to.